src/main/scala/com/cloudera/sparkts/DateTimeIndex.scala

```diff
@@ -261,14 +359,280 @@ class IrregularDateTimeIndex(
+    islice(range.head, range.last + 1)
+  }
+
+  override def islice(start: Int, end: Int): HybridDateTimeIndex = {
+    require(start < end, s"start($start) should be less than end($end)")
+
+    val startIndex = binarySearch(0, indices.length - 1, start)
+    val endIndex = binarySearch(0, indices.length - 1, end)
+
+    val alignedStart = start - sizeOnLeft(startIndex)
+    val alignedEnd = end - sizeOnLeft(endIndex)
+
+    var newIndices: Array[DateTimeIndex] = Array.empty
+
+    newIndices =
+      if (startIndex == endIndex) {
+        Array(indices(startIndex).islice(alignedStart, alignedEnd))
+      } else {
+        val startDateTimeIndex = indices(startIndex)
+        val endDateTimeIndex = indices(endIndex)
+
+        val startSlice = Array(startDateTimeIndex.islice(alignedStart, startDateTimeIndex.size))
+        val endSlice = Array(endDateTimeIndex.islice(0, alignedEnd))
+
+        if (endIndex - startIndex == 1)
+          startSlice ++ endSlice
+        else
+          startSlice ++ indices.slice(startIndex + 1, endIndex) ++ endSlice
+      }
+
+    new HybridDateTimeIndex(newIndices, dateTimeZone)
+  }
+
+  override def first: ZonedDateTime = indices(0).first.withZoneSameInstant(dateTimeZone)
+
+  override def last: ZonedDateTime = indices(indices.length - 1).last
+    .withZoneSameInstant(dateTimeZone)
+
+  override def zone: ZoneId = dateTimeZone
+
+  override def size: Int = indices.map(_.size).sum
+
+  override def dateTimeAtLoc(loc: Int): ZonedDateTime = {
+    val i = binarySearch(0, indices.length - 1, loc)
+    if (i > -1) indices(i).dateTimeAtLoc(loc - sizeOnLeft(i))
+    else throw new ArrayIndexOutOfBoundsException(s"no dateTime at loc $loc")
+  }
+
+  private def binarySearch(low: Int, high: Int, i: Int): Int = {
+    if (low <= high) {
+      val mid = (low + high) >>> 1
+      val midIndex = indices(mid)
+      val sizeOnLeftOfMid = sizeOnLeft(mid)
+      if (i < sizeOnLeftOfMid) binarySearch(low, mid - 1, i)
+      else if (i >= sizeOnLeftOfMid + midIndex.size) binarySearch(mid + 1, high, i)
+      else mid
+    } else -1
+  }
+
+  override def locAtDateTime(dt: ZonedDateTime): Int = {
+    val i = binarySearch(0, indices.length - 1, dt)._1
+    if (i > -1) {
+      val loc = indices(i).locAtDateTime(dt)
+      if (loc > -1) sizeOnLeft(i) + loc
+      else -1
+    }
+    else -1
+  }
+
+  override def locAtDateTime(dt: Long): Int = {
+    locAtDateTime(longToZonedDateTime(dt, dateTimeZone))
+  }
+
+  override def insertionLoc(dt: ZonedDateTime): Int = {
+    val loc = binarySearch(0, indices.length - 1, dt)._2
+    if (loc >= 0) {
+      sizeOnLeft(loc) + indices(loc).insertionLoc(dt)
+    } else if (dt.isBefore(first)) {
+      0
+    } else {
+      size
+    }
+  }
+
+  override def insertionLoc(dt: Long): Int = {
+    insertionLoc(longToZonedDateTime(dt, dateTimeZone))
+  }
+
+  private def binarySearch(low: Int, high: Int, dt: ZonedDateTime): (Int, Int) = {
```

Comment:
Can you include a header comment here to indicate what the two tuple values mean?